kore/src/SMT.hs

```diff
@@ -123,63 +143,213 @@ different threads may be interleaved; use 'inNewScope' to acquire exclusive
 access to the solver for a sequence of commands.
 
  -}
-newtype SMT a = SMT { getSMT :: ReaderT (MVar Solver) IO a }
-    deriving (Applicative, Functor, Monad)
+newtype SMT a = SMT { getSMT :: ReaderT Environment IO a }
+    deriving
+        ( Applicative
+        , Functor
+        , Monad
+        , MonadCatch
+        , MonadIO
+        , MonadThrow
+        , Reader.MonadReader Environment
```

Comment:
I think deriving this instance is causing the problems you expect. We should *not* provide this instance. `Environment` is an implementation detail; this would violate encapsulation.